returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,253 @@
+"""
+Provides :class:`PostprocessingDataset`.
+"""
+
+from __future__ import annotations
+
+from numpy import ndarray
+from typing import Any, Callable, Dict, Iterator, List, Optional, Tuple, Union
+
+from returnn.datasets.basic import DatasetSeq
+from returnn.datasets.util.vocabulary import Vocabulary
+from returnn.tensor import Tensor, TensorDict
+from returnn.tensor.dim import Dim
+from returnn.tensor.tensor import ShapeMismatchException
+from .basic import Dataset, init_dataset
+from .cached2 import CachedDataset2
+from .util.strings import str_to_numpy_array
+
+__all__ = ["PostprocessingDataset"]
+
+
+class PostprocessingDataset(CachedDataset2):
+    """
+    A dataset that allows for generic post-processing of data from another dataset
+    using a function on the segment level and on the level of multiply segments via
+    an iterator.
+
+    This allows integrating various data augmentation techniques like e.g. Mixup, or
+    speed perturbation into the data loading pipeline.
+
+    The integration into the data loading pipeline makes it easy to distribute the
+    data processing work across multiple CPU cores using `MultiProcDataset` and in
+    turn frees the GPU from data preprocessing tasks.
+
+    Both functions can be specified at the same time, and their results will be
+    combined by first applying the segment-level function and then the
+    multiple-segments-level function.
+
+    Example usage:
+        from returnn.tensor.dim import Dim, DimTypes
+
+        time_dim = Dim(None, kind=DimTypes.Spatial)
+        new_data_dim = Dim(128)
+
+        train = {
+            "class": "PostprocessingDataset",
+            "dataset": {
+                "class": "HDFDataset",
+                "files": ["/path/to/data.hdf"],
+            },
+            # at least one of them:
+            "map_seq": map_seq,  # (data: TensorDict) -> TensorDict
+            "map_seq_stream": map_seqs,  # (iter: Iterator[TensorDict]) -> Iterator[TensorDict]
+            # only required when data shapes change wrt. the wrapped dataset:
+            "map_outputs": {
+                "data": { "dims": [time_dim, new_data_dim] },
```

Comment:
Always use properly PEP8 formatted code in examples.
```suggestion
                "data": {"dims": [time_dim, new_data_dim]},
```